js/commonjs/rewrite_tests_for_commonjs.js

```diff
@@ -0,0 +1,69 @@
+/**
+ * @fileoverview Utility to translate test files to CommonJS imports.
+ *
+ * This is a somewhat hacky tool designed to do one very specific thing.
+ * All of the test files in *_test.js are written with Closure-style
+ * imports (goog.require()).  This works great for running the tests
+ * against Closure-style generated code, but we also want to run the
+ * tests against CommonJS-style generated code without having to fork
+ * the tests.
+ *
+ * Closure-style imports import each individual type by name.  This is
+ * very different than CommonJS imports which are by file.  So we put
+ * special comments in these tests like:
+ *
+ * // CommonJS-LoadFromFile: test_pb
+ * goog.require('proto.jspb.test.CloneExtension');
+ * goog.require('proto.jspb.test.Complex');
+ * goog.require('proto.jspb.test.DefaultValues');
+ *
+ * This script parses that special comment and uses it to generate proper
+ * CommonJS require() statements so that the tests can run and pass using
+ * CommonJS imports.
+ */
+
+var lineReader = require('readline').createInterface({
+  input: process.stdin,
+  output: process.stdout
+});
+
+function tryStripPrefix(str, prefix) {
+  if (str.lastIndexOf(prefix) !== 0) {
+    throw "String: " + str + " didn't start with: " + prefix;
+  }
+  return str.substr(prefix.length);
+}
+
+var module = null;
+var pkg = null;
+lineReader.on('line', function(line) {
+  var is_require = line.match(/goog\.require\('([^']*)'\)/);
+  var is_loadfromfile = line.match(/CommonJS-LoadFromFile: ([^ ]*) (.*)/);
+  var is_settestonly = line.match(/goog.setTestOnly()/);
+  if (is_settestonly) {
+    // Remove this line.
+  } else if (is_require) {
+    if (module) {  // Skip goog.require() lines before the first directive.
+      var full_sym = is_require[1];
+      var sym = tryStripPrefix(full_sym, pkg);
+      console.log("google_protobuf.exportSymbol('" + full_sym + "', " + module + sym + ', global);');
+    }
+  } else if (is_loadfromfile) {
+    if (!module) {
+      console.log("var google_protobuf = require('google-protobuf');");
+      console.log("var asserts = require('closure_asserts_commonjs');");
+      console.log("var global = Function('return this')();");
+      console.log("");
+      console.log("// Bring asserts into the global namespace.");
+      console.log("google_protobuf.object.extend(global, asserts);");
```

Comment:
The overall goal of this change is to provide a simple out-of-the-box experience for people who are not using Closure Compiler or Closure-style require/provide.

Leaving `goog.require()` in the source forces users to generate `deps.js` and make sure all the paths line up. If `goog.require()` fails, they have to dig into this unfamiliar system and figure out why. I went through the pain of this when setting up our Closure-based tests, and had to write this somewhat hacky script to help Closure find files in the current directory: https://github.com/google/protobuf/blob/master/js/node_loader.js. (This was when I was already using `goog/bootstrap/nodejs.js`, see: https://github.com/google/protobuf/blob/master/js/jasmine.json#L8).

With the approach in this CL, the entire protobuf runtime library gets compiled to a single `google-protobuf.js` that does not depend on any `deps.js` or external libraries. This also makes things more convenient for browsers: they just need one `<script>` tag for the protobuf runtime library.

The future is ES6 imports anyway, which are file-oriented. I think moving away from symbol-based includes and towards file-based includes like `require` is moving things in the right direction for convenience and for the future. At this point I see `goog.require()` and `deps.js` as a legacy setup, and I don't want to push a legacy setup on our users.